File: server/src/main/com/broll/gainea/server/core/objects/Unit.java
package com.broll.gainea.server.core.objects;

import com.broll.gainea.net.NT_Unit;
import com.broll.gainea.server.core.battle.BattleContext;
import com.broll.gainea.server.core.battle.BattleResult;
import com.broll.gainea.server.core.battle.FightingPower;
import com.broll.gainea.server.core.objects.buffs.BuffableInt;
import com.broll.gainea.server.core.objects.buffs.IntBuff;
import com.broll.gainea.server.core.player.Player;

public abstract class Unit extends MapObject {

    private BuffableInt<MapObject> maxHealth = new BuffableInt<>(this, 0);
    private BuffableInt<MapObject> power = new BuffableInt<>(this, 0);
    private BuffableInt<MapObject> health = new BuffableInt<>(this, 0);
    private BuffableInt<MapObject> attacksPerTurn = new BuffableInt<>(this, 1); //default 1 attack

    private BuffableInt<MapObject> numberPlus = new BuffableInt<>(this, 0);

    private int attackCount;
    private boolean moveOrAttackRestriction = true; //usually units can only attack or move in one turn
    private boolean attacked = false;
    private boolean moved = false;
    protected boolean controllable = true;
    private int type = NT_Unit.TYPE_MALE;
    private int kills;
    private boolean justSpawned = true;

    public Unit(Player owner) {
        super(owner);
        health.setMinValue(0);
        maxHealth.setMinValue(1);
        power.setMinValue(0);
    }

    public static void copy(Unit from, Unit to) {
        to.maxHealth = from.maxHealth.copy(to);
        to.power = from.power.copy(to);
        to.health = from.health.copy(to);
        to.owner = from.owner;
        to.attackCount = from.attackCount;
        to.moveCount = from.moveCount;
        to.attacksPerTurn = from.attacksPerTurn.copy(to);
        to.movesPerTurn = from.movesPerTurn.copy(to);
        to.numberPlus = from.numberPlus.copy(to);
        to.setMoveOrAttackRestriction(from.isMoveOrAttackRestriction());
        to.setIcon(from.getIcon());
        to.setLocation(from.getLocation());
        to.setName(from.getName());
        to.setScale(from.getScale());
    }

    @Override
    public void turnStart() {
        super.turnStart();
        justSpawned = false;
        attackCount = 0;
        moved = false;
        attacked = false;
    }

    @Override
    public boolean hasRemainingMove() {
        if (justSpawned) return false;
        if (moveOrAttackRestriction && attacked) {
            return false;
        }
        return super.hasRemainingMove();
    }

    public boolean hasRemainingAttack() {
        if (justSpawned) return false;
        if (moveOrAttackRestriction && moved) {
            return false;
        }
        return attackCount < attacksPerTurn.getValue();
    }

    public void onDeath(BattleResult throughBattle) {

    }

    public FightingPower calcFightingPower(BattleContext context) {
        return new FightingPower(this).changeNumberPlus(numberPlus.getValue());
    }

    @Override
    public void moved() {
        super.moved();
        moved = true;
    }

    public void attacked() {
        attackCount++;
        attacked = true;
    }

    public void setStats(int power, int health) {
        setPower(power);
        setHealth(health);
    }

    public boolean isMoveOrAttackRestriction() {
        return moveOrAttackRestriction;
    }

    public void setMoveOrAttackRestriction(boolean moveOrAttackRestriction) {
        this.moveOrAttackRestriction = moveOrAttackRestriction;
    }

    @Override
    public BuffableInt<MapObject> getMovesPerTurn() {
        return super.getMovesPerTurn();
    }

    public Player getOwner() {
        return owner;
    }

    public boolean takeDamage() {
        return takeDamage(1);
    }

    public boolean takeDamage(int damage) {
        boolean aliveBefore = isAlive();
        health.addValue(-damage);
        return aliveBefore && isDead();
    }

    public void heal(int heal) {
        health.addValue(heal);
        if (health.getValue() > maxHealth.getValue()) {
            heal();
        }
    }

    public boolean isDead() {
        return health.getValue() <= 0;
    }

    public boolean isAlive() {
        return !isDead();
    }

    public void heal() {
        health.setValue(maxHealth.getRootValue());
    }

    public void setHealth(int health) {
        this.health.setValue(health);
        this.maxHealth.setValue(health);
    }

    public void changeHealth(int change) {
        this.health.addValue(change);
        this.maxHealth.addValue(change);
    }

    public void setPower(int power) {
        this.power.setValue(power);
    }

    public BuffableInt<MapObject> getPower() {
        return power;
    }

    public void addHealthBuff(IntBuff buff) {
        health.addBuff(buff);
        maxHealth.addBuff(buff);
    }

    public void clearBuffs() {
        power.clearBuffs();
        health.clearBuffs();
        maxHealth.clearBuffs();
        attacksPerTurn.clearBuffs();
        movesPerTurn.clearBuffs();
    }

    public BuffableInt<MapObject> getHealth() {
        return health;
    }

    public BuffableInt<MapObject> getMaxHealth() {
        return maxHealth;
    }

    public BuffableInt<MapObject> getAttacksPerTurn() {
        return attacksPerTurn;
    }

    public BuffableInt<MapObject> getNumberPlus() {
        return numberPlus;
    }

    public void setOwner(Player owner) {
        this.owner = owner;
    }

    public int getBattleStrength() {
        return getPower().getValue() + getHealth().getValue();
    }

    @Override
    public NT_Unit nt() {
        NT_Unit unit = new NT_Unit();
        fillBattleObject(unit);
        return unit;
    }

    protected void fillBattleObject(NT_Unit unit) {
        fillObject(unit);
        unit.health = health.getValue().shortValue();
        unit.maxHealth = maxHealth.getValue().shortValue();
        unit.power = power.getValue().shortValue();
        unit.type = (byte) type;
        unit.kills = (short) kills;
        if (owner != null) {
            unit.owner = (short) owner.getServerPlayer().getId();
        }
    }

    public boolean isHurt() {
        return health.getValue() < maxHealth.getValue();
    }

    public void setType(int type) {
        this.type = type;
    }

    public int getKills() {
        return kills;
    }

    public void addKill() {
        this.kills++;
    }

    public boolean isControllable() {
        return controllable;
    }

    @Override
    public String toString() {
        return "Unit{" +
                "id=" + getId() +
                ", name='" + getName() +
                ", " + getPower() + "/" + getHealth() +
                ", location=" + getLocation() + '\'' +
                '}';
    }
}

File: server/src/main/com/broll/gainea/server/core/objects/Unit.kt
package com.broll.gainea.server.core.objects

import com.broll.gainea.net.NT_Unit
import com.broll.gainea.server.core.battle.BattleContext
import com.broll.gainea.server.core.battle.BattleResult
import com.broll.gainea.server.core.battle.FightingPower
import com.broll.gainea.server.core.objects.buffs.BuffableInt
import com.broll.gainea.server.core.objects.buffs.IntBuff
import com.broll.gainea.server.core.player.Player

abstract class Unit(owner: Player) : MapObject(owner), IUnit {
    override var maxHealth: BuffableInt<MapObject> = BuffableInt(this, 0)

    override var power: BuffableInt<MapObject> = BuffableInt(this, 0)

    override var health: BuffableInt<MapObject> = BuffableInt(this, 0)

    override var attacksPerTurn: BuffableInt<MapObject> = BuffableInt(this, 1) //default 1 attack

    override var numberPlus: BuffableInt<MapObject> = BuffableInt(this, 0)

    private var attackCount = 0
    var isMoveOrAttackRestriction = true //usually units can only attack or move in one turn
    private var attacked = false
    private var moved = false
    var controllable = true
    private var type = NT_Unit.TYPE_MALE.toInt()
    override var kills = 0
    private var justSpawned = true

    init {
        health.setMinValue(0)
        maxHealth.setMinValue(1)
        power.setMinValue(0)
    }

    override fun prepareForTurnStart() {
        super.prepareForTurnStart()
        justSpawned = false
        attackCount = 0
        moved = false
        attacked = false
    }

    override fun hasRemainingMove(): Boolean {
        if (justSpawned) return false
        return if (isMoveOrAttackRestriction && attacked) {
            false
        } else super.hasRemainingMove()
    }

    fun hasRemainingAttack(): Boolean {
        if (justSpawned) return false
        return if (isMoveOrAttackRestriction && moved) {
            false
        } else attackCount < attacksPerTurn.value
    }

    open fun onDeath(throughBattle: BattleResult?) {}
    open fun calcFightingPower(context: BattleContext) =
        FightingPower(this).changeNumberPlus(numberPlus.value)


    override fun moved(fromPlayerAction: Boolean) {
        super.moved(fromPlayerAction)
        if (fromPlayerAction) {
            moved = true
        }
    }

    fun attacked(fromPlayerAction: Boolean = false) {
        if (fromPlayerAction) {
            attackCount++
            attacked = true
        }
    }

    fun setStats(power: Int, health: Int) {
        setPower(power)
        setHealth(health)
    }


    fun takeDamage(damage: Int = 1): Boolean {
        val aliveBefore = alive
        health.addValue(-damage)
        return aliveBefore && dead
    }

    fun heal(heal: Int) {
        health.addValue(heal)
        if (health.value > maxHealth.value) {
            health.value = maxHealth.value
        }
    }

    fun heal() {
        health.value = maxHealth.value
    }

    fun setHealth(health: Int) {
        this.health.value = health
        maxHealth.value = health
    }

    fun addHealth(change: Int) {
        health.addValue(change)
        maxHealth.addValue(change)
    }

    fun overwriteHealth(health: BuffableInt<MapObject>) {
        this.health = health
    }

    fun setPower(power: Int) {
        this.power.value = power
    }

    fun addHealthBuff(buff: IntBuff) {
        health.addBuff(buff)
        maxHealth.addBuff(buff)
    }

    fun clearBuffs() {
        power.clearBuffs()
        health.clearBuffs()
        maxHealth.clearBuffs()
        attacksPerTurn.clearBuffs()
        movesPerTurn.clearBuffs()
    }

    val battleStrength: Int
        get() = power.value + health.value

    override fun nt(): NT_Unit {
        val unit = NT_Unit()
        fillBattleObject(unit)
        return unit
    }

    protected fun fillBattleObject(unit: NT_Unit) {
        fillObject(unit)
        unit.health = health.value.toShort()
        unit.maxHealth = maxHealth.value.toShort()
        unit.power = power.value.toShort()
        unit.type = type.toByte()
        unit.kills = kills.toShort()
    }

    val hurt: Boolean
        get() = health.value < maxHealth.value

    fun setType(type: Int) {
        this.type = type
    }

    open fun killedEnemy(unit: Unit) {
        kills++
    }

    override fun toString(): String {
        return "Unit{" +
                "id=" + id +
                ", name='" + name +
                ", " + power + "/" + health +
                ", location=" + location + '\'' +
                '}'
    }

    fun rank() = power.value + health.value

    companion object {
        fun copy(from: Unit, to: Unit) {
            to.maxHealth = from.maxHealth.copy(to)
            to.power = from.power.copy(to)
            to.health = from.health.copy(to)
            to.owner = from.owner
            to.attackCount = from.attackCount
            to.moveCount = from.moveCount
            to.attacksPerTurn = from.attacksPerTurn.copy(to)
            to.movesPerTurn = from.movesPerTurn.copy(to)
            to.numberPlus = from.numberPlus.copy(to)
            to.isMoveOrAttackRestriction = from.isMoveOrAttackRestriction
            to.icon = from.icon
            to.location = from.location
            to.name = from.name
            to.scale = from.scale
        }
    }
}
